vavr/src/main/java/io/vavr/collection/Iterator.java

```diff
@@ -2144,22 +2145,27 @@ public T getNext() {
 
         @Override
         public boolean hasNext() {
-            while (next == null && that.hasNext()) {
+            return nextDefined || searchNext();
+        }
+
+        private boolean searchNext() {
+            while (that.hasNext()) {
                 final T elem = that.next();
                 final U key = keyExtractor.apply(elem);
                 if (!known.contains(key)) {
                     known = known.add(key);
+                    nextDefined = true;
                     next = elem;
+                    return true;
                 }
             }
-            return next != null;
+            return false;
         }
 
         @Override
         public T getNext() {
-            final T result = next;
-            next = null;
-            return result;
+            nextDefined = false;
+            return next;
```

Comment:
We still hold the reference on the last consumed `next` value.
Corner case: a reference to this iterator remains. Then the last iterator value that was consumed can't be garbage collected.

I think I will set `next = null` in another commit to be clean here.